File: Blind_75_List/DP - Multidimensional/72. Edit Distance.cpp
//Problem Link: https://leetcode.com/problems/edit-distance/

class Solution {
public:
    
    int minDistance(string word1, string word2) {
        int row = word2.size(), col = word1.size();
        vector<vector<int>> dp(row+1, vector<int>(col + 1, 0));
        for(int i = 0; i<=row; i++){
            dp[i][col] =  row - i;
        }
        for(int j = 0; j<=col; j++){
            dp[row][j] =  col - j;
        }
        for(int i = row-1; i>=0; i--){
            for(int j = col-1; j>=0; j--){
                if (word1[j] == word2[i]){
                    dp[i][j] = dp[i+1][j+1];
                }
                else{
                    dp[i][j] = 1 + min({
                        dp[i+1][j],
                        dp[i][j+1],
                        dp[i+1][j+1]
                    });
                }
                
            }
        }
    
        return dp[0][0];
        
    }
};

File: Blind_75_List/DP - Multidimensional/72. Edit Distance.py
# Problem Link: https://leetcode.com/problems/edit-distance/

class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        dp = {("","") : 0}
        def lev(a: str,b:str)->int:
            if dp.get((a,b), -1) != -1:
                return dp[(a,b)]
            
            if len(a) == 0:
                dp[(a,b)] = len(b)
                return len(b)
            
            if len(b) == 0:
                dp[(a,b)] = len(a)
                return len(a)
            
            if a[0] == b[0]:
                dp[(a,b)] = lev(a[1:], b[1:])
                return dp[(a,b)]
            
            x = dp[(a,b[1:])] = lev(a,b[1:])
            
            y = dp[(a[1:],b[1:])] = lev(a[1:],b[1:])
            
            z = dp[(a[1:],b)] = lev(a[1:],b)
            
            dp[(a,b)] = min(x,y,z) + 1
            return dp[(a,b)]
        return lev(word1,word2)